src/Expander.hs

```diff
@@ -983,24 +981,36 @@ runTask (tid, localData, task) = withLocal localData $ do
       case value of
         ValueSyntax syntax -> do
           forkExpandSyntax dest syntax
-        other -> expandEval $ evalErrorType "syntax" other
+        other -> do
+          p <- currentPhase
+          throwError $ MacroEvaluationError p $ evalErrorType "syntax" other
     ContinueMacroAction dest value (closure:kont) -> do
-      result <- expandEval $ apply closure value
-      case result of
-        ValueMacroAction macroAction -> do
-          forkInterpretMacroAction dest macroAction kont
-        other -> expandEval $ evalErrorType "macro action" other
+      case apply closure value of
+        Left err -> do
+          p <- currentPhase
+          throwError
+            $ MacroEvaluationError p
+            $ evalErrorType "macro action"
+            $ erroneousValue
+            $ projectError err
```

Comment:
this error-handling logic is repeated several time, wasn't it nicer when `expandEval` was running in the `Expand` monad and was thus able to throw this error right away?

Reply:
I'm hesitant to DRY it out because I still want to workshop the design of the debugger and how it hooks into `expandEval`.

> throw this error right away

This action will be calling the debugger. Immediately exiting the debugger should then be throwing the error and reporting the error back to the user. Perhaps `Expand` will have to be a monad transformer over `DebugM`? I'm honestly not sure if anyone has made a CommonLisp style debugger in Haskell yet.

Reply:
Now that I understand what `erroneousValue` does, I don't think this code is correct. It claims that the problem is that there is a value which should be a macro action, but its actual type is something else. But that problem is raised below, on L1001:
```
throwError $ MacroEvaluationError p $ evalErrorType "macro action" other
```

The problem on _this_ line is some unknown thing which happened during evaluation, not necessarily a type error. Instead of using `err` to construct a type error, we should propagate `err` upwards as-is.